### PR TITLE
Spellbook and Spell progression mechanics

### PR DESCRIPTION
**Pull request by hunharibo**
First iteration of the spellbook and spell progression mechanic.

* Personal spellbook to view the details of known spells and spellcasting stats.
* Spellcasting stats (max winds of magic and recharge rate) depend on Intelligence skill
* Spell trainers in towns who stay in the keeps with full dialog branch for 
* 1, learning new spells
* 2, testing for magical affinity for non-spellcasters
* 3, specializing in a certain advanced lore after minor magic (which is universally accessible for spellcasters)
* 4, advancing in spellcasting level (minor->entry->adept->master) to get access to higher tier spells
* a rudimentary basic quest system to add a condition for the advancing in casting level, right now it just requires to cast a spell for condition, but Jax and lore guys are thinking about more complex quest objectives (defeat certain NPC etc.)
* 2 new minor magic spells for testing, Minor Heal (self cast single target heal) and Ward of Arrows (self cast single target physical resistance). These are actually from the spell list document on the drive prepared by the lore/game design guys, so don't get hung up on the name of it :D

### CODE REVIEW

File: CSharpSourceCode/Abilities/Ability.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ public abstract class Ability : IDisposable
         public AbilityScript AbilityScript { get; private set; }
 
         public AbilityCrosshair Crosshair { get; private set; }
-
+        public virtual AbilityEffectType GetAbilityEffectType => Template.AbilityEffectType;
```
Comment (Badulundabad) [inline CSharpSourceCode/Abilities/Ability.cs]: i guees you just forgot to rename it to AbilityEffectType

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs [hunk 2]
```diff
@@ -94,6 +104,14 @@ private void HandleAnimations()
         internal void OnCastComplete()
         {
             if(CurrentState == AbilityModeState.Casting) _currentState = AbilityModeState.Idle;
+            if(Game.Current.GameType is Campaign)
+            {
+                var quest = Campaign.Current.QuestManager.Quests.FirstOrDefault(x => x.GetType() == typeof(AdvanceSpellCastingLevelQuest)) as AdvanceSpellCastingLevelQuest;
```
Comment (Badulundabad) [inline CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs]: minor, but just for readability i would change to ...`x => x is AdvanceSpellCastingLevelQuest`...

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs [hunk 3]
```diff
@@ -69,6 +70,15 @@ public override void OnAgentHit(Agent affectedAgent, Agent affectorAgent, int da
...
+            if (!_hasInitializedForMainAgent)
+            {
+                if(Agent.Main != null)
+                {
+                    SetUpCastStanceParticles();
+                    _hasInitializedForMainAgent = true;
+                }
+            }
+
             if (IsAbilityModeAvailableForMainAgent())
```
Comment (Badulundabad) [inline CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs]: minor. i would write here 'else' and remove _hasInitializedForMainAgent from IsAbilityModeAvailableForMainAgent(). Yeah it's war against double checks)

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs [hunk 4]
```diff
@@ -317,10 +338,14 @@ public SummonedCombatant GetSummoningCombatant(Team team)
 
         protected override void OnAgentControllerChanged(Agent agent, Agent.ControllerType oldController)
         {
-            if (agent.Controller != Agent.ControllerType.Player || Agent.Main == null || !Agent.Main.IsAbilityUser())
+            if (agent.Controller == Agent.ControllerType.Player)
             {
-                return;
+                _hasInitializedForMainAgent = false;
             }
+        }
```
Comment (Badulundabad) [inline CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs:343]: what use of this method now? `_hasInitializedForMainAgent` is false by default

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs [hunk 5]
```diff
@@ -329,9 +354,8 @@ protected override void OnAgentControllerChanged(Agent agent, Agent.ControllerTy
                 _psys[0] = TOWParticleSystem.ApplyParticleToAgentBone(Agent.Main, _castingStanceParticleName, Game.Current.HumanMonster.MainHandItemBoneIndex, out entity);
                 _psys[1] = TOWParticleSystem.ApplyParticleToAgentBone(Agent.Main, _castingStanceParticleName, Game.Current.HumanMonster.OffHandItemBoneIndex, out entity);
                 EnableCastStanceParticles(false);
-                _currentState = AbilityModeState.Off;
-                _isAbilityUser = true;
             }
+            _currentState = AbilityModeState.Off;
```
Comment (Badulundabad) [inline CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs]: minor. i would move this to class initialization, like value by default

File: CSharpSourceCode/Abilities/SpellBook/SpellBookMapIconCampaignBehaviour.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            Layer = mapView.GauntletLayer;
+            _layer = Layer as GauntletLayer;
+            _movie = _layer.LoadMovie("SpellBookMapIcon", _vm);
+        }
+
+        protected override void OnFinalize()
+        {
+            _vm.OnFinalize();
+            _vm = null;
+            _layer.ReleaseMovie(_movie);
```
Comment (Badulundabad) [inline CSharpSourceCode/Abilities/SpellBook/SpellBookMapIconCampaignBehaviour.cs:64]: good, i didn't know there is such method, i think would be nice to do it with all views. Not in this pr

File: CSharpSourceCode/Abilities/SpellBook/SpellBookMapIconCampaignBehaviour.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+using TaleWorlds.Library;
+
+namespace TOW_Core.Abilities.SpellBook
+{
+    public class SpellBookMapIconCampaignBehaviour : CampaignBehaviorBase
+    {
+        private bool _initialized;
+        public override void RegisterEvents()
+        {
+            CampaignEvents.TickEvent.AddNonSerializedListener(this, Initialize);
```
Comment (Badulundabad) [inline CSharpSourceCode/Abilities/SpellBook/SpellBookMapIconCampaignBehaviour.cs]: why TickEvent? can we use something that fires once per session?

File: CSharpSourceCode/Abilities/Scripts/AbilityScript.cs [hunk 8]
```diff
@@ -182,17 +182,11 @@ protected void TriggerEffect(Vec3 position, Vec3 normal)
...
-                if (_ability.Template.CrosshairType == CrosshairType.Targeted)
+                if(_ability.Template.AbilityTargetType == AbilityTargetType.Self)
                 {
-                    var crosshair = (TargetedCrosshair)_ability.Crosshair;
-                    var target = crosshair.LastTargetIndex != -1 ? Mission.Current.Agents.FirstOrDefault(a => a.Index == crosshair.LastTargetIndex) : null;
-                    if (target != null)
-                    {
-                        effect.Trigger(position, normal, _casterAgent, new List<Agent>(1) { target });
-                    }
-                    return;
```
Comment (Badulundabad) [inline CSharpSourceCode/Abilities/Scripts/AbilityScript.cs:193]: why did you remove passing of target for targeted crosshair?

File: CSharpSourceCode/Battle/AttributeSystem/CustomBattleMoraleModel/TOWCampaignBattleMoraleModel.cs [hunk 9]
```diff
@@ -5,23 +5,20 @@
 using TOW_Core.Abilities;
 using TOW_Core.Utilities.Extensions;
 
-namespace TOW_Core.Battle.ObjectDataExtensions.CustomBattleMoralModel
+namespace TOW_Core.Battle.AttributeSystem.CustomBattleMoralModel
 {
-    public class CustomBattleMoralModel
+    public class TOWCampaignBattleMoraleModel : SandboxBattleMoraleModel
```
Comment (Badulundabad) [inline CSharpSourceCode/Battle/AttributeSystem/CustomBattleMoraleModel/TOWCampaignBattleMoraleModel.cs]: minor. Should be TOR not TOW i guess

**Review by Badulundabad — CHANGES_REQUESTED** [review]
So much new cool stuff, great work! 
I left mostly minor comments, please take a look. And finally, why did you remove assimilation behavior and another related stuff? Also it isn't tied with pr theme and moreover was not mentioned in description of it.

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs (hunk 4 above)
Comment (hunharibo) [reply CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs:343]: Yes, but if the main agent changes - there are some mods that allow this, to take control of a "normal" troop, then we can support that by setting the flag to false Again, so we handle the changing of the main agent with initialization. This is mainly just a precaution for players who would want to use our mod together with other mods.

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs (hunk 2 above)
Comment (hunharibo) [reply CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs]: agree, will change. Made a static method for this on the class, but forgot to change this occurance to that. Thanks for noticing.

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs (hunk 5 above)
Comment (hunharibo) [reply CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs]: will do

File: CSharpSourceCode/Abilities/Scripts/AbilityScript.cs (hunk 8 above)
Comment (hunharibo) [reply CSharpSourceCode/Abilities/Scripts/AbilityScript.cs:193]: because it crashes for AI. AI has no crosshair and they also need to be able to cast single target spells. Test it yourself. Custom battle with enemy lord being empire lord. As soon as they cast Dart, the game crashes.

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs (hunk 4 above)
Comment (Badulundabad) [reply CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs:343]: You're right

File: CSharpSourceCode/Abilities/Scripts/AbilityScript.cs (hunk 8 above)
Comment (Badulundabad) [reply CSharpSourceCode/Abilities/Scripts/AbilityScript.cs:193]: Ugh, I thought I managed this case

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs (hunk 3 above)
Comment (hunharibo) [reply CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs]: Modified accordingly.

File: CSharpSourceCode/Battle/AttributeSystem/CustomBattleMoraleModel/TOWCampaignBattleMoraleModel.cs (hunk 9 above)
Comment (hunharibo) [reply CSharpSourceCode/Battle/AttributeSystem/CustomBattleMoraleModel/TOWCampaignBattleMoraleModel.cs]: Renamed all models that were still TOW for consistency.

Comment (Badulundabad) [reply CSharpSourceCode/Battle/AttributeSystem/CustomBattleMoraleModel/TOWCampaignBattleMoraleModel.cs]: fantastic)

File: CSharpSourceCode/Abilities/SpellBook/SpellBookMapIconCampaignBehaviour.cs (hunk 7 above)
Comment (hunharibo) [reply CSharpSourceCode/Abilities/SpellBook/SpellBookMapIconCampaignBehaviour.cs]: changed to PushedScreen event as discussed on discord

Comment (Badulundabad) [reply CSharpSourceCode/Abilities/SpellBook/SpellBookMapIconCampaignBehaviour.cs]: awesome

**Review by Badulundabad — APPROVED** [review]
Nicely done! Approve

File: CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs [hunk 10]
```diff
@@ -94,6 +103,14 @@ private void HandleAnimations()
         internal void OnCastComplete()
         {
             if(CurrentState == AbilityModeState.Casting) _currentState = AbilityModeState.Idle;
+            if(Game.Current.GameType is Campaign)
+            {
+                var quest = AdvanceSpellCastingLevelQuest.GetCurrentActiveIfExists();
```
Comment (Z3rca) [inline CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs:108]: later down the line, we should move the whole quest logic, to some kind of Event System. But that's obviously something for later

File: CSharpSourceCode/CampaignSupport/TownBehaviours/SpellTrainerInTownBehaviour.cs [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        }
+
+        private void OnSessionLaunched(CampaignGameStarter obj)
+        {
+            AddDialogs(obj);
+        }
+
+        private void AddDialogs(CampaignGameStarter obj)
+        {
+            obj.AddDialogLine("trainer_start", "start", "choices", "What do you want?", spelltrainerstartcondition, null, 200, null);
```
Comment (Z3rca) [inline CSharpSourceCode/CampaignSupport/TownBehaviours/SpellTrainerInTownBehaviour.cs:50]: would like if we can later move the strings to an XML file that can be edited without the need to jump into the code. now this would be overkill

**Review by Z3rca — APPROVED** [review]
comments are more or less stuff for future... no problems with this. 👍
